test/vtl.test.ts

```diff
@@ -333,18 +333,61 @@ test("chain map over list", () =>
         .map((item) => `hello ${item}`);
     }),
     `#set($v1 = [])
+#foreach($item in $context.arguments.list)
+#set($item = "hello \${item}")
+#set($v2 = "hello \${item}")
+$util.qr($v1.add($v2))
+#end
+${returnExpr("$v1")}`
+  ));
+
+test("chain map over list multiple array", () =>
+  testCase(
+    reflect((context: AppsyncContext<{ list: string[] }>) => {
+      return context.arguments.list
+        .map((item, _i, _arr) => `hello ${item}`)
+        .map((item, _i, _arr) => `hello ${item}`);
+    }),
+    `#set($v1 = [])
 #set($v2 = [])
 #foreach($item in $context.arguments.list)
+#set($_i = $foreach.index)
+#set($_arr = $context.arguments.list)
 #set($v3 = "hello \${item}")
 $util.qr($v2.add($v3))
 #end
 #foreach($item in $v2)
+#set($_i = $foreach.index)
+#set($_arr = $v2)
 #set($v4 = "hello \${item}")
 $util.qr($v1.add($v4))
 #end
 ${returnExpr("$v1")}`
   ));
 
+test("chain map over list complex", () =>
+  testCase(
+    reflect((context: AppsyncContext<{ list: string[] }>) => {
+      return context.arguments.list
+        .map((item, i, arr) => {
+          const x = i + 1;
+          return `hello ${item} ${x} ${arr.length}`;
+        })
+        .map((item2, ii) => `hello ${item2} ${ii}`);
```

Comment:
what happens if you name this `i` instead of `ii`?

Reply:
Will it conflict with the other chain calls?

Reply:
It's fine because 1. It would be the same value, map can't change i 2. It would overwrite the previous one, 3. The ts compiler will ensure there isn't a conflict.

We may be able to optimize later by proving that a set doesn't make a meaningful change, but just a non functional optimization.

Reply:
The previous test shows what happens.